Dagr overhaul v1

nh13 opened the pull request.

nh13 commented on api/shared/src/main/scala/dagr/api/models/TaskInfo.scala:
```diff
@@ -0,0 +1,153 @@
+trait TaskInfoLike  {
+  def id         : Option[TaskId]
+  def attempts   : Int
+  def script     : Option[String]
```
`script` and `log` must be a string since `java.nio.file.Path` is not available for `scala-js`.

tfenne replied:
That's really shitty.  I don't like the idea that our internal model objects have to be restricted because of the use of scala JS to provide a UI.  Is there some way that `TaskInfo` could still have `Path`s, and also expose them as `String`s for the JS layer?

tfenne replied:
I think too... if we really can't fix this, then they need better names, as it's no longer clear that `script: Option[String]` is the path to a script, not the contents of a script.

tfenne replied:
Coming back to this after reading a bunch more ... so it seems like you have a bunch of code to convert some of our objects to JS and back again.  Is there no way to register a simple handler for Path to do likewise?

nh13 commented on api/shared/src/main/scala/dagr/api/DagrApi.scala:
```diff
@@ -0,0 +1,130 @@
+}
+
+// NB: this trait cannot extend anything, as required by autowire
+trait DagrApi {
```
This is implemented on the server-side and called by the scala-js side (via [autowire](https://github.com/lihaoyi/autowire))

nh13 commented on api/shared/src/main/scala/dagr/api/models/DagrApiDef.scala:
```diff
@@ -0,0 +1,119 @@
+
+import upickle.Js
+
+object DagrApiDef {
```
Perhaps a better name would be `DagrApiPickling`

tfenne replied:
Or `DagrJsConversions`?  Sucks that you have to do this.

tfenne replied:
I'm already feeling like a broken record .. but some scaladoc would be really nice!

nh13 commented on api/shared/src/main/scala/dagr/api/models/TaskInfo.scala:
```diff
@@ -0,0 +1,153 @@
+
+/** The execution information for a task.  Used for extrenal read-only access to [[TaskInfo]]. Any execution system
+  * should extend [[TaskInfo]] instead class to store their specific metadata. */
+trait TaskInfoLike  {
```
A trait since extending a case class just feels wrong.

tfenne replied:
Agreed .. but then your comment about suggests to do exactly that ;)

nh13 commented on api/shared/src/main/scala/dagr/api/models/TaskInfo.scala:
```diff
@@ -0,0 +1,153 @@
+object TaskInfo {
+  import DagrApiDef._
+
+  implicit val query2Writer: Writer[TaskInfo] = Writer[TaskInfo] {
```
You have to put these implicit `val`s in the companion object for pickling with `upickle`.

tfenne replied:
The amount of scala-js stuff invading the rest of dagr makes me really sad.  I was hoping it would be nearly totally isolated to a `ui` package and not affect the internals at all.

nh13 commented on api/shared/src/main/scala/dagr/api/DagrApi.scala:
```diff
@@ -0,0 +1,130 @@
+  def statuses(): Seq[TaskStatus]
+
+  /** The method a server implementation should implement. */
+  protected def _query(name: Seq[String] = Seq.empty, // any name pattern
```
Only here so we can convert `String` to `ResourceSet`.  I don't think this is necessary anymore since I moved `ResourceSet` into the `dagr.api` package.

tfenne replied:
That would be nice.  So many params to repeat.

nh13 commented on api/shared/src/main/scala/dagr/api/models/TaskInfoResponse.scala:
```diff
@@ -0,0 +1,68 @@
+import upickle.Js
+import upickle.default.{Reader, Writer}
+
+case class TaskInfoResponse(infos: Seq[TaskInfo], report: Option[String] = None)
```
The response to the `DagrApi.query`  method.

nh13 commented on api/shared/src/main/scala/dagr/api/models/TaskStatus.scala:
```diff
@@ -0,0 +1,75 @@
+object TaskStatus {
+  /** Builds a [[TaskStatus]] from the string.  The returned status will be an anonymous sub-class. */
+  /*
+  def parse(s: String): TaskStatus = {
```
Not needed by the API, but could be useful.

nh13 commented on api/shared/src/main/scala/dagr/api/models/TimePoint.scala:
```diff
@@ -0,0 +1,68 @@
+        throw new IllegalArgumentException(s"Could not parse TimePoint '$s'")
+    }
+  }
+  */
```
Not needed, remove.

nh13 commented on core/src/main/scala/dagr/core/cmdline/DagrCoreMain.scala:
```diff
@@ -103,11 +109,15 @@ class DagrCoreArgs(
-  var interactive: Boolean = false
+  var interactive: Boolean = false,
+  @arg(doc = "Use the experimental execution system.")
+  val experimentalExecution: Boolean = false,
```
These changes include the later `execsystem2` changes as it was a PITA to break it up.

nh13 commented on core/src/main/scala/dagr/core/exec/Executor.scala:
```diff
@@ -0,0 +1,124 @@
+            failFast: Boolean = false
+           )(implicit ex: ExecutionContext): Executor = {
+    if (experimentalExecution) {
+      GraphExecutor(new LocalTaskExecutor(systemResources=resources, scriptsDirectory=scriptsDirectory, logDirectory=logDirectory))
```
the new execution system creeping in.

nh13 commented on core/src/main/scala/dagr/core/execsystem/GraphNode.scala:
```diff
@@ -48,7 +48,7 @@ class GraphNode(var task: Task,
 
   def taskId: TaskId = taskInfo.taskId
-  def taskInfo: TaskExecutionInfo = task.taskInfo
+  def taskInfo: TaskExecutionInfo = task.execsystemTaskInfo
```
So ugly, but I have to make it work somehow.

nh13 commented on core/src/main/scala/dagr/core/tasksystem/Task.scala:
```diff
@@ -23,20 +23,167 @@
+    }
+
+    def scriptPath: Option[FilePath] = this.script.map(PathUtil.pathTo(_))
+    def logPath: Option[FilePath]    = this.script.map(PathUtil.pathTo(_))
```
Only needed because `script` is a type `Option[String]`.  I could make a `lazy va`l.

nh13 commented on core/src/main/scala/dagr/core/reporting/ExecutionLogger.scala:
```diff
@@ -0,0 +1,211 @@
+
+import scala.collection.mutable
+
+object ExecutionLogger {
```
This is the class that logs the execution state and relationships between tasks.  The file produced is used as input to the replay system.

tfenne replied:
Maybe we could call it `ReplayLogger` then?

nh13 commented on core/src/main/scala/dagr/core/reporting/ExecutionLogger.scala:
```diff
@@ -0,0 +1,211 @@
+    }
+  }
+
+  object Relationship {
```
Relationship is redundant as we can generate it from the definition, but it's a sanity check nonetheless.  Happy to get rid of it.

tfenne replied:
I think I would kill it.  It's an opportunity to introduce inconsistency when editing the replay log.

nh13 commented on core/src/main/scala/dagr/core/reporting/TaskStatusLogger.scala:
```diff
@@ -0,0 +1,40 @@
+import dagr.core.tasksystem.Task.TaskInfoLike
+
+/** A simple logger that delegates to [[dagr.core.tasksystem.Task.TaskInfo#logTaskMessage]]. */
+class TaskStatusLogger extends TaskLogger {
```
This is the log you see on the terminal.

nh13 commented on core/src/main/scala/dagr/core/reporting/TopLikeStatusReporter.scala:
```diff
@@ -0,0 +1,327 @@
+  def eraseToEOL: String = "\033[K"
+}
+
+object TopLikeStatusReporter {
```
This is the output you would see in interactive mode.

nh13 commented on core/src/main/scala/dagr/core/reporting/TopLikeStatusReporter.scala:
```diff
@@ -0,0 +1,327 @@
+}
+
+
+/** A top-like status reporter that prints execution information to the terminal. The
```
This is _mostly_ old code, with some updates to the API.

nh13 replied:
Note, in a latter commit, I added a lot more tests, including one for the interactive top-like reporter.

nh13 commented on core/src/main/scala/dagr/core/tasksystem/Task.scala:
```diff
@@ -23,20 +23,167 @@
 /** Utility methods to aid in working with a task. */
 object Task {
 
+  trait TaskInfoLike extends dagr.api.models.TaskInfoLike with Ordered[TaskInfoLike] {
```
The base trait for `TaskInfo`s for `dagr.core`. Added to make it ordered, add in both a `scriptPath` and `logPath` method, and finally to store a reference to the actual task (good luck getting the latter to compile in scala-js).

nh13 commented on core/src/main/scala/dagr/core/tasksystem/Task.scala:
```diff
@@ -23,20 +23,167 @@
+    * @param exitCode the exit-code for the task, if any
+    * @param throwable a throwable generated during execution, if any
+    */
+  private[dagr] abstract class TaskInfo
```
The `TaskInfo` for the execution systems.  Updated so we can have various callbacks handled internally.

nh13 commented on core/src/main/scala/dagr/core/tasksystem/Task.scala:
```diff
@@ -23,20 +23,167 @@
+    private[core] def status_=(status: TaskStatus): Instant = {
+      val instant = Instant.now
+      update(status, instant)
+      this.task._executor.foreach(_.record(info=this))
```
A callback to all things that want to know when the status changes for a task.

nh13 commented on core/src/main/scala/dagr/core/tasksystem/Task.scala:
```diff
@@ -204,6 +362,21 @@ trait Task extends Dependable {
+  final def make(): Traversable[_ <: Task] = {
+    val tasks = getTasks.toSeq
+    tasks.foreach(_._executor = this._executor)
+    this._executor.foreach(_.register(this, tasks:_*))
```
A callback to all things that want to know when a task is first seen/registered.

nh13 commented on core/src/main/scala/dagr/core/tasksystem/Task.scala:
```diff
@@ -23,20 +23,167 @@
+    protected[core] def startDate: Option[Instant]
+
+    /** The instant that the task finished executing. */
+    protected[core] def endDate: Option[Instant]
```
We can get rid of these methods when we dump the old exec system.

nh13 commented on core/src/main/scala/dagr/core/tasksystem/Task.scala:
```diff
@@ -141,20 +288,31 @@ object Task {
+  private[core] def taskInfo_=(info: TaskInfo) = {
+    this._taskInfo = Some(info)
+    this._executor.foreach(_.record(info))
+  }
```
This is ugly, sorry.

nh13 commented on core/src/main/scala/dagr/core/tasksystem/Task.scala:
```diff
@@ -141,20 +288,31 @@ object Task {
+  private[core] def execsystemTaskInfo : TaskExecutionInfo = this._taskInfo.getOrElse(unreachable(s"Task info should be defined for task '$name'")) match {
+    case info: TaskExecutionInfo => info
+    case info => throw new IllegalStateException(s"For task info, expected type 'TaskExecutionInfo' but found '${info.getClass.getSimpleName}")
+  }
```
This is ugly too.

nh13 commented:
@tfenne I tried splitting all the changes I have made into discrete commits so you can review one-by-one, and then after each, I can rebase and fixup.  The first one up is: https://github.com/fulcrumgenomics/dagr/pull/306/commits/7c50a9e057b2f234dc8e998ebce309600b67325d

tfenne commented on api/shared/src/main/scala/dagr/api/DagrApi.scala:
```diff
@@ -0,0 +1,130 @@
+import dagr.api.DagrApi.TaskId
+import dagr.api.models._
+
+/*
```
Can this commented out code go?

tfenne commented on api/shared/src/main/scala/dagr/api/DagrApi.scala:
```diff
@@ -0,0 +1,130 @@
+}
+*/
+
+trait DagrApiConfiguration {
```
Some documentation of these traits and objets would be helpful since I don't yet understand why they're here.

tfenne commented on api/shared/src/main/scala/dagr/api/DagrApi.scala:
```diff
@@ -0,0 +1,130 @@
+trait DagrApi {
+
+
+  final def query(name: Seq[String] = Seq.empty, // any name pattern
```
Also needs better documentation.  I'm guessing that all of these things are optional, but it's not totally clear.

tfenne commented on api/shared/src/main/scala/dagr/api/DagrApi.scala:
```diff
@@ -0,0 +1,130 @@
+}
+
+// NB: this trait cannot extend anything, as required by autowire
+trait DagrApi {
```
Unless this represents the whole public API of dagr, I don't love calling it `DagrApi`.  E.g. if I wanted to embed dagr in another app to run workflows, this name makes me think I would configure dagr and submit workflows through this same API.

Perhaps `DagrStatusApi` or `DagrQueryApi` or something along those lines?

nh13 replied:
I don't think it's unreasonable to think we will have APIs for those, though I can understand perhaps wanting a different trait for each.

tfenne commented on api/shared/src/main/scala/dagr/api/models/DagrApiDef.scala:
```diff
@@ -0,0 +1,119 @@
+
+object DagrApiDef {
+
+  val MissingValue: String = ""
```
Should this be private?

tfenne commented on api/shared/src/main/scala/dagr/api/models/Resource.scala:
```diff
@@ -20,28 +20,13 @@
-
-  /** The heap size of the JVM. */
-  val heapSize: Memory = new Memory(Runtime.getRuntime.maxMemory)
+trait ResourceParsing {
```
Can you please explain why this had to get pulled out of the `Resource` object into a separate trait?

nh13 replied:
Both `dagr.core.exec.Resource`  and `dagr.api.models.Resource` extend it.  The former contains methods for querying and manipulating system resources.  The reason that this can't be bundled into the API, is that `scala-js` doesn't support a lot of the Java libraries when it compiles all these for use in Javascript.  For example, reflection is right out, and any dependency (like `oshi`) needs to be separately compiled for `scala-js`.

tfenne commented on api/shared/src/main/scala/dagr/api/models/ResourceSet.scala:
```diff
@@ -20,14 +20,18 @@
 
   val empty = ResourceSet(0, 0)
+  val Inf = ResourceSet(Double.MaxValue, Long.MaxValue)
+  @deprecated("use `Inf` instead", since="0.1.3")
```
Why the change?  Also, then wouldn't it be better to have `val infinite = Inf`?

nh13 replied:
There was a reason, but since I didn't document it, I am going to revert.

tfenne commented on api/shared/src/main/scala/dagr/api/models/TaskInfo.scala:
```diff
@@ -0,0 +1,153 @@
+import scala.collection.mutable.ListBuffer
+import java.time.Instant
+
+/** The execution information for a task.  Used for extrenal read-only access to [[TaskInfo]]. Any execution system
```
extrenal -> external

tfenne commented on api/shared/src/main/scala/dagr/api/models/TaskInfo.scala:
```diff
@@ -0,0 +1,153 @@
+
+/** The execution information for a task.  Used for extrenal read-only access to [[TaskInfo]]. Any execution system
+  * should extend [[TaskInfo]] instead class to store their specific metadata. */
+trait TaskInfoLike  {
```
How about `TaskInfoView` since it's a read only view?

tfenne commented on api/shared/src/main/scala/dagr/api/models/TaskInfo.scala:
```diff
@@ -0,0 +1,153 @@
+  def timePoints :  Traversable[TimePoint]
+
+  /** The instant the task reached the current status. */
+  def statusTime : Instant
```
Are `status` and `statusTime` really needed?  Or at least, could they be implemented in the trait as `timePoints.last.status` or similar?

tfenne commented on api/shared/src/main/scala/dagr/api/models/TaskInfo.scala:
```diff
@@ -0,0 +1,153 @@
+object TaskInfo {
+  import DagrApiDef._
+
+  implicit val query2Writer: Writer[TaskInfo] = Writer[TaskInfo] {
```
Should this be `taskInfoToWriter`?

tfenne commented on api/shared/src/main/scala/dagr/api/models/TaskInfo.scala:
```diff
@@ -0,0 +1,153 @@
+      Js.Obj(tuples:_*)
+  }
+
+  implicit val query2Reader: Reader[TaskInfo] = Reader[TaskInfo] {
```
`taskInfoToReader`?

tfenne commented on core/src/main/scala/dagr/core/DagrDef.scala:
```diff
@@ -34,4 +34,10 @@ package dagr.core
 object DagrDef {
   /** The type of identifier used to uniquely identify tasks tracked by the execution system. */
   type TaskId = BigInt
```
I had no recollection we were using a `BigInt` here.  Do you think a simple `Int` is insufficient?  I imagine many other things will break before we hit 2e9 tasks.

tfenne commented on core/src/main/scala/dagr/core/exec/Executor.scala:
```diff
@@ -0,0 +1,124 @@
+  /** Create a new executor. */
+  def apply(experimentalExecution: Boolean,
+            resources: SystemResources,
+            scriptsDirectory: Option[DirPath] = None,
```
scripts is pluralized by log isn't, please pick one.

tfenne commented on core/src/main/scala/dagr/core/exec/Executor.scala:
```diff
@@ -0,0 +1,124 @@
+trait Executor extends FinalStatusReporter {
+
+  /** A list of [[TaskCache]] to use to determine if a task should be manually succeeded. */
+  protected val taskCaches: ListBuffer[TaskCache] = ListBuffer[TaskCache]()
```
These should all probably be typed as `Buffer` not `ListBuffer`.

tfenne commented on core/src/main/scala/dagr/core/exec/Executor.scala:
```diff
@@ -0,0 +1,124 @@
+  /** The loggers to be notified when a task's status is updated. */
+  private val _loggers: ListBuffer[TaskLogger] = ListBuffer[TaskLogger](new TaskStatusLogger)
+
+  /** The loggers to be notified when a task's status is updated. */
```
Copy/pasted comment.

tfenne commented on core/src/main/scala/dagr/core/exec/Executor.scala:
```diff
@@ -0,0 +1,124 @@
+    this._loggers.foreach(_.record(info=info))
+  }
+
+  /** The method that will be called on the result of `Task.getTasks`. */
```
Please define in terms of what it does, not by whom/when it is called.

tfenne commented on core/src/main/scala/dagr/core/exec/Executor.scala:
```diff
@@ -0,0 +1,124 @@
+  }
+
+  /** Adds the [[dagr.core.reporting.ReportingDef.TaskLogger]] to the list of loggers to be notified when a task's status is updated. */
+  private def withLogger(logger: TaskLogger): Unit = this.synchronized {
```
Methods named `withX` should generally return a new/updated version of `this` with X added.  I.e. method return "this, with X".  Otherwise just call them "addX".  I think for these I would recommend just calling the methods `+=`, so you can write `this += someNewLogger`

tfenne replied:
And now I see the `with` methods below do return `this`, just this one doesn't for some reason.

tfenne commented on core/src/main/scala/dagr/core/exec/Executor.scala:
```diff
@@ -0,0 +1,124 @@
+  def statusFrom(ordinal: Int): TaskStatus
+
+  /** The list of statuses ordered by ordinal */
+  def statuses: Seq[TaskStatus]
```
If you make this return `IndexedSeq` you could implement `statusFrom` as `statuses(ordinal)`.

tfenne commented on core/src/main/scala/dagr/core/exec/Executor.scala:
```diff
@@ -0,0 +1,124 @@
+  /** Returns the log directory. */
+  def logDir: DirPath
+
+  /** Start the execution of this task and all tasks that depend on it.  Returns the number of tasks that were not
```
I'm confused.  It says "Start the execution", but then the return implies that this method blocks during execution.  Which is correct?

tfenne commented on core/src/main/scala/dagr/core/exec/Resource.scala:
```diff
@@ -0,0 +1,82 @@
+/** The resources needed for an execution system */
+object SystemResources {
+  /** Creates a new SystemResources that is a copy of an existing one. */
+  def apply(that: SystemResources): SystemResources = {
```
Do we even need this?  Wouldn't `that.copy()` inlined be the same thing?

nh13 replied:
Hey I am just copy and pasting here.

tfenne commented on core/src/main/scala/dagr/core/exec/TaskCache.scala:
```diff
@@ -0,0 +1,323 @@
+
+import scala.collection.mutable
+
+/** All implementations that decide if a given task should be executed should implement this. */
```
Needs a lot more documentation as to what it's purpose is and how it fits into the bigger picture.

tfenne commented on core/src/main/scala/dagr/core/exec/TaskCache.scala:
```diff
@@ -0,0 +1,323 @@
+  /** Returns true if the task should be executed, false otherwise. */
+  def execute(task: Task): Boolean
+
+  /** The method that will be called to registered the children of the parent.  If no children are given, then the task
```
registere[d]

tfenne commented on core/src/main/scala/dagr/core/exec/TaskCache.scala:
```diff
@@ -0,0 +1,323 @@
+    * is the root of all tasks.  If there is one child and it is equal to the parent, then it is a unit task and
+    * should have already been registered, and so nothing is done.  Otherwise, the children are individually registered.
+    * */
+  def register(parent: Task, child: Task*): Unit = {
```
The contract for this method is really confusing.  Does it really make a big difference to efficiency to allow registration of multiple tasks in one go?  It's confusing to me because it's not clear which tasks are being registered, and why you have to have so many `if`s below.  Why not simply:
`register(task: Task, parent: Option[Task])`
which would make it crystal clear that `task` is the thing being registered, and it may optionally have a parent?

tfenne commented on core/src/main/scala/dagr/core/exec/TaskCache.scala:
```diff
@@ -0,0 +1,323 @@
+  protected def taskSet: Set[Task]
+
+  /** Returns true if the task is a root task (has never depended on a task).*/
+  protected def isRoot(parent: Task, child: Task*): Boolean = taskSet.isEmpty && child.isEmpty
```
This requires that the root is the first registered task, but that's not explained anywhere.

tfenne commented on core/src/main/scala/dagr/core/exec/TaskCache.scala:
```diff
@@ -0,0 +1,323 @@
+  private[exec] def tasks: Seq[Task] = this.taskSet.toSeq
+
+  /** Returns the tasks that have been registered. */
+  protected def taskSet: Set[Task]
```
If this is the one that's used everywhere, why not call this `tasks` instead of `taskSet` and call the tasks accessor for testing `_taskSeq` so it's not easily confused?

tfenne commented on core/src/main/scala/dagr/core/exec/TaskCache.scala:
```diff
@@ -0,0 +1,323 @@
+}
+
+/** A very inefficient class to determine if tasks should be executed upon replay */
+class SimpleTaskCache(replayLogLines: Seq[String], source: Option[String] = None) extends TaskCache {
```
Why is source a String and not Path?

tfenne commented on core/src/main/scala/dagr/core/exec/TaskCache.scala:
```diff
@@ -0,0 +1,323 @@
+/** A very inefficient class to determine if tasks should be executed upon replay */
+class SimpleTaskCache(replayLogLines: Seq[String], source: Option[String] = None) extends TaskCache {
+
+  private val replayLog: String = source.getOrElse("Unknown")
```
What is this used for?  I'm assuming just logging/toStringing since it can't be relied on to not be "Unknown".  Perhaps `sourceString` would be better?

tfenne commented on core/src/main/scala/dagr/core/exec/TaskCache.scala:
```diff
@@ -0,0 +1,323 @@
+
+  // TODO: are relationships even needed if we have both the definition of child, who has the parent code?
+  /** Definitions, relationships, and statuses in the replay log */
+  private val (definitions, relationships, statuses) = {
```
Somewhere, perhaps in the class-level scaladoc, the replay log format needs to be explained.

tfenne commented on core/src/main/scala/dagr/core/exec/TaskCache.scala:
```diff
@@ -0,0 +1,323 @@
+      }
+    // Check the parent and child in each relationship have a definition
+    this.relationships.flatMap { r => Seq(r.parentCode, r.childCode) }
+      .find { code => !this.definitions.exists(_.code == code) }
```
This is n^2 or worse ... could be slow on large logs.

tfenne commented on core/src/main/scala/dagr/core/exec/TaskCache.scala:
```diff
@@ -0,0 +1,323 @@
+      }
+    // Check that a task if is found for each status
+    this.statuses.map { s => s.definitionCode }
+      .find { code => !this.definitions.exists(_.code == code) }
```
Ditto here on the n^2.  Maybe just make a temporary set of `this.definitions.map(_.code).toSet` or similar?

tfenne commented on core/src/main/scala/dagr/core/execsystem/TaskExecutionInfo.scala:
```diff
@@ -23,45 +23,37 @@
+                        script: FilePath,
+                        log: FilePath,
+                        resources: Option[ResourceSet] = Some(ResourceSet(0, 0)))
+  extends Task.TaskInfo(task=task, initStatus=initStatus, id=Some(initId), script=Some(script.toString), log=Some(log.toString), resources=resources)
```
I'm thoroughly confused by the fact that there are both `TaskInfo` classes specific to each exec system, and a parent one in `Task` and another one in `dagr.api.models`.  How do the one in `Task` and the one in `models` relate to one another?

tfenne commented on core/src/main/scala/dagr/core/execsystem/TaskStatus.scala:
```diff
@@ -21,42 +21,71 @@
-  def isTaskDone(taskStatus: TaskStatus, failedIsDone: Boolean = true): Boolean = {
-    (isTaskFailed(taskStatus) && failedIsDone) || SUCCEEDED == taskStatus || MANUALLY_SUCCEEDED == taskStatus || STOPPED == taskStatus
+  def done(taskStatus: RootTaskStatus, failedIsDone: Boolean = true): Boolean = {
+    val completed = taskStatus.isInstanceOf[Completed]
```
I think this should either take `TaskStatus` or `require(taskStatus.isInstanceOf[TaskStatus]` because it will do the wrong thing if somehow handed the wrong type.

nh13 replied:
Agreed.

tfenne commented on core/src/main/scala/dagr/core/execsystem/TaskStatus.scala:
```diff
@@ -21,42 +21,71 @@
+    * @param taskStatus the status of the task
+    * @return true if the task has failed, false otherwise
+    */
+  def failed(taskStatus: RootTaskStatus): Boolean = taskStatus.isInstanceOf[Failed]
```
Ditto here.  Can't this take `TaskStatus`?

tfenne commented on core/src/main/scala/dagr/core/execsystem/TopLikeStatusReporter.scala:
```diff
@@ -20,275 +20,46 @@
-    print("\n")
-  }
+                            protected val loggerOut: Option[ByteArrayOutputStream] = None,
+                            protected val print: String => Unit = print) extends BaseTopLikeStatusReporter {
```
Instead of having execsystem like sub-classes, is there not some way to just take in the exec system or query API?

tfenne commented on core/src/main/scala/dagr/core/execsystem2/DependencyGraph.scala:
```diff
@@ -0,0 +1,145 @@
+/**
+  * A trait that classes that track and update dependencies on tasks should implement.
+  */
+trait DependencyGraph {
```
This trait needs a lot more documentation.  The fact that there are zero methods for adding _dependencies_ between tasks implies that all of the dependency creation is done when adding tasks, but looking at their upstream and downstream dependencies.  But that's never actually mentioned.

Also, is there any special handling of, e.g. Pipelines? I'm guessing that within this trait and it's implementations that explicit dependencies (`a ==> b`) and parent child (i.e. pipeline A contains task B) are handled the same way, but none of that is documented.

tfenne commented on core/src/main/scala/dagr/core/execsystem2/DependencyGraph.scala:
```diff
@@ -0,0 +1,145 @@
+trait DependencyGraph {
+
+  /** Add a task to the dependency graph and returns true if the task has no dependencies, false otherwise.  */
+  def add(task: Task): Boolean
```
Needs a lot more explanation.  Is the `add` method required to navigate the task's dependencies and add those relationships?  Transitively back to the root?  What about tasks that depend on this task?

tfenne replied:
Also, can a task acquire dependencies later?

tfenne commented on core/src/main/scala/dagr/core/execsystem2/DependencyGraph.scala:
```diff
@@ -0,0 +1,145 @@
+  /** Removes this task as a dependency for all other tasks in this dependency graph.  The task should not depend on
+    * any tasks, and all tasks that depend on it will have their dependency on this task removed.
+    */
+  def remove(task: Task): Seq[Task]
```
Does it also remove the `task`?  I.e. will `contains(task) == false` after calling this method?
What happens if the `task` still has dependencies?  Does it do nothing or throw an exception?
What are the returned tasks?  I'm guessing it's the set of tasks that no longer have any dependencies after this task was removed??

tfenne commented on core/src/main/scala/dagr/core/execsystem2/DependencyGraph.scala:
```diff
@@ -0,0 +1,145 @@
+    * information.
+    * @param task
+    */
+  def exceptIfCyclicalDependency(task: Task): Unit
```
Why does this take in a task?
Should this be `requireNoCycles: Unit`?

tfenne commented on core/src/main/scala/dagr/core/execsystem2/DependencyGraph.scala:
```diff
@@ -0,0 +1,145 @@
+}
+
+/**
+  * A very simple dependency graph that uses a [[CountDownLatch]] on the number of dependencies for a [[Task]] to
```
Doesn't actually use a count down latch.  `Latch` is used to mean a lock which releases or flips.  You're using a _counter_, not a `CountDownLatch`.

tfenne commented on core/src/main/scala/dagr/core/execsystem2/DependencyGraph.scala:
```diff
@@ -0,0 +1,145 @@
+  * A very simple dependency graph that uses a [[CountDownLatch]] on the number of dependencies for a [[Task]] to
+  * block until a task has no dependencies.
+  */
+private class SimpleDependencyGraph extends DependencyGraph with LazyLogging {
```
I'm not sure what to make of this.  On the one hand I never liked the old system of duplicating all the task relationships into a separate graph.  On the other hand, calling these classes `DependencyGraph` strongly implies to me that's what you're doing.  Maybe it's as simple as renaming these to something else that doesn't imply it's storing a graph separate from the task relationships?

Also, if we're going to rely on `task.tasksDependedOn` then I wonder why even keep a separate counter and not just use things like `task.tasksDependedOn.filterNot(_.status.succeeded).count` instead?  That way we won't have any issues if the tasks set of dependencies is fiddled with post-registration.

tfenne commented on core/src/main/scala/dagr/core/execsystem2/DependencyGraph.scala:
```diff
@@ -0,0 +1,145 @@
+    * dependencies.
+    */
+  def remove(task: Task): Seq[Task] = {
+    require(task.tasksDependedOn.isEmpty,
```
This is weird to me, that we use a counter to track how many dependencies a task has, but then rely on removing the task as a dependency from downstream tasks to check that the task has no dependencies before removing.  I think this class should either be totally reliant on Task for it's information or totally reliant on it's own internal state, but not a mixture of the two.

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+import scala.concurrent.{Await, ExecutionContext, Future, blocking}
+import scala.util.{Failure, Success}
+
+// TODO: support failFast
```
Yikes - is this still a todo?

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+// TODO: support failFast
+
+/** Coordinates between the dependency graph ([[DependencyGraph]]) and task executor ([[TaskExecutor]]) given a (root)
+  * task to execute.
```
... to do what?

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+trait GraphExecutor[T<:Task] extends Executor {
+  /** Start the execution of this task and all tasks that depend on it.  Returns the number of tasks that were not
+    * executed.  A given task should only be attempted once. */
+  def execute(rootTask: Task): Int
```
Should be much clearer that it's expected to block until there is no more work that can be done.
Is it thread safe?  I.e. can multiple calls to `execute()` be handled concurrently?
A task should be attempted only once?  Even if it implements `Retry`?  Why?

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+    new GraphExecutorImpl(taskExecutor=taskExecutor, dependencyGraph=DependencyGraph())
+}
+
+/** An implementation of an executor of a tasks that have dependencies.
```
a tasks -> tasks

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+  private val _tasks: mutable.Set[Task] = ExecDef.concurrentSet()
+
+  /** A lock to synchronize when the task execution or dependency information is updated. */
+  private val lock: Object = (dependencyGraph, _tasks)
```
Why not just lock on `this` since that would seem more natural?  Also `(dependencyGraph, _tasks)` seems to imply that the identity of the lock object has meaning, but using `(dependencyGraph, _tasks)` is no different that using `new Object()`.

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+  /** Returns the log directory. */
+  def logDir: DirPath = this.taskExecutor.logDir
+
+  /** Start the execution of this task and all tasks that depend on it.  Returns the number of tasks that were not
```
Use `@return` to describe the return.

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+    }
+
+    // Return the number of tasks known not to succeed
+    this._tasks.map(_.taskInfo).count { info => !info.status.success }
```
This seems problematic.  If multiple tasks are running concurrently then this will give the wrong answer.  If it's not allowed to run multiple root tasks concurrently, then the API should prevent that.

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+      // wait until all tasks that depended on this task are processed
+      Future.sequence(dependentFutures) map { _: Seq[Task] => task }
+    }
+  }
```
I think it's preferable to avoid the pattern of ignoring the thing you're mapping over, e.g. the flatMap above.  Since you don't care about the returns of `registerTask` why not do something like:

```scala
Future {
  lockIt { task.tasksDependingOnThisTask.foreach(registerTask) }
  buildAndExecuteTask(task)
  this.dependencyGraph.remove(t)
} map { ts =>
  Future.sequence(ts.map(processTask)) map { _: Seq[Task] => task }
}
```

I'm sure I've screwed something up, but essentially, why not have the first three parts (register, execute, remove dependencies) in one Future, and then map to create the child futures and sequence them?

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+
+  /** Builds a task and proceeds based on if it create other tasks or itself should be executed.  Completes when
+    * itself and any children have executed.  It does not update the dependency graph. */
+  private def buildAndExecute(parent: Task): Future[Task] = failFutureWithTaskStatus(parent) { // tag any failures during build or execution
```
If it creates children tasks, why aren't these entered into the dependency graph??

tfenne replied:
Ok, so the scaladoc is a lie!  This method will call `executeMultiTask()` which does in fact update the dependency graph.

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+          executeWithTaskExecutor(parent)
+        }
+        else {
+          updateMetadata(parent, ManuallySucceeded)
```
I don't think ManuallySucceeded is right here.  It should either be Succeeded, or perhaps a replay-specific status like `SucceededPreviously`?

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+
+    buildTask(parent) flatMap {
+      case x :: Nil if x == parent => // one task and it returned itself, so execute it
+        requireNoDependencies(parent)
```
Why is it necessary to repeat this here?

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+
+  /** Build the task and catch any exceptions during the call to [[Task.make()]]. */
+  private def buildTask(task: Task): Future[Seq[Task]] = failWithFailedToBuild(task) {
+    requireNoDependencies(task)
```
Is `requireNoDependencies(task)` your favorite method?  It's called so many times!

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+  private def buildTask(task: Task): Future[Seq[Task]] = failWithFailedToBuild(task) {
+    requireNoDependencies(task)
+    val tasks = task.make().toList
+    if (tasks.isEmpty) throw new IllegalArgumentException(s"No tasks built from task: '${task.name}'")
```
It might be nice to support a task returning no tasks, since that would eliminate a lot of the need for NoOp tasks.  E.g.:
```scala
class MaybeTask(task: Task, when => Boolean) extends Task {
  def build = if (when) Seq(task) else Seq.empty()
}
```

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+      updateMetadata(parent, Running)
+
+      // For each child task, first add them to the dependency graph.  Next, add any dependent tasks (task that depend
+      // on the child task), to the graph.  The parent task can complete once they have all been added to the
```
I don't understand/agree with that last sentence.  Sure it can only complete once all the children have completed?

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+          childTasks.foreach { child =>
+            require(!this.dependencyGraph.contains(child), s"child '${child.name}' of parent '${parent.name}' already in the graph")
+          }
+          val result = childTasks flatMap { child =>
```
result -> immediatelyExecutableChildren

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+          }
+          // check for cyclical dependencies since a new sub-tree has been added
+          this.dependencyGraph.exceptIfCyclicalDependency(parent)
+          // NB: returns true if the task has no dependencies, false otherwise
```
What returns true?  You're returning executable child tasks aren't you?

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+      // other multitude of reasons.
+      // TODO: partial functions here
+      this.taskExecutor.execute(t)
+      // TODO: could have a status for 'Scheduled' (eg. Submitted -> Scheduled -> Running)
```
Yes please.

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+  private def executeWithTaskExecutor(task: Task): Future[T] = {
+    Future { task.asInstanceOf[T] } flatMap { t: T =>
+      val subFuture     = submissionFuture   (t)
+      val execFuture    = executionFuture    (t, subFuture=subFuture)
```
I wonder if it would be better to replace this with handing the TaskExecutor a `() => Unit` that it executes when the task transitions to running?  Rather than having extra futures?

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+  }
+
+  /** Provides synchronization and signals that this may block. */
+  private def lockIt[A](body: =>A): A = blocking { this.lock.synchronized(body) }
```
lockIt -> `locked` or `locking`?

tfenne commented on core/src/main/scala/dagr/core/execsystem2/GraphExecutor.scala:
```diff
@@ -0,0 +1,416 @@
+
+  /** Ensure that the task has no dependencies.  If missingOk is set, then if do not throw an exception if the task
+    * is not known. */
+  private def requireNoDependencies(task: Task, missingOk: Boolean = false): Unit = {
```
`missingOk` doesn't appear to ever be set to true, can it be removed?

tfenne commented on core/src/main/scala/dagr/core/execsystem2/TaskExecutor.scala:
```diff
@@ -0,0 +1,67 @@
+  * of things not specified for various executors (ex. SGE, PBS, local, mixed) on purpose. */
+trait TaskExecutor[T<:Task] {
+  /** simple name (not unique) for the executor. */
+  def name: String = getClass.getSimpleName
```
Do you want to do your santizing trick here?

tfenne commented on core/src/main/scala/dagr/core/execsystem2/TaskExecutor.scala:
```diff
@@ -0,0 +1,67 @@
+  def execute(task: T): Future[Future[T]]
+
+  /** terminate a task, returns true if successful, false otherwise, None if it knows nothing about the task. */
+  def kill(task: T, duration: Duration = Duration.Zero): Option[Boolean]
```
There's a lot of use of `Option[Boolean]` here and elsewhere.  Two thoughts:
1) Do we really need to support all these cases of "it's ok to pass me semi-invalid input"?
2) Maybe it would be better to have some simple sealed enums for these cases that can be matched on?  E.g. `enum { Killed, StillAlive, WtfThatsNotATaskOfMine }`

tfenne commented on core/src/main/scala/dagr/core/execsystem2/TaskInfo.scala:
```diff
@@ -0,0 +1,45 @@
+  extends Task.TaskInfo(task=task, initStatus=initStatus) {
+
+  /** Gets the instant that the task was submitted to the execution system. */
+  override protected[core] def submissionDate: Option[Instant] = this(Pending)
```
Having `TaskInfo.apply(TaskStatus)` return an `Option` feels wrong, since it's normal for Map-like things to return X from apply() and Option[X] from get().

tfenne commented on core/src/main/scala/dagr/core/execsystem2/TaskStatus.scala:
```diff
@@ -0,0 +1,69 @@
+sealed abstract class TaskStatus extends IntEnumEntry with RootTaskStatus {
+  override def ordinal = this.value
+  /** Returns true if this status indicates any type of success, false otherwise. */
+  def success: Boolean = this.isInstanceOf[TaskStatus.Succeeded]
```
Why not just define all three methods as returning `false` and then selectively override below?

nh13 replied:
To my, and this totally could be a preference, it is much cleaner and concise to have them defined here, versus having each task status spell each of these three out.

tfenne commented on core/src/main/scala/dagr/core/execsystem2/TaskStatus.scala:
```diff
@@ -0,0 +1,69 @@
+
+  /** Trait for all statuses during execution */
+  sealed trait Executing         extends TaskStatus
+  case object Submitted          extends Executing     { val description: String = "Submitted for execution";                val value: Int = 2 }
```
Remind me how this is different than Queued?

nh13 replied:
For pipelines
    
    No dependencies -> `Queued` -> build the task -> `Running`

For task executed by the underling executor (ex. bash)
  
    No dependencies -> `Queued` -> build the task-> `Submitted` -> give to the underlying executor -> `Running`.

The "give to the underlying executor" could fail for a number of reasons including but not limited to the task requires more resources than the system has, the task fail while executing, etc.

tfenne commented on core/src/main/scala/dagr/core/execsystem2/local/LocalTaskExecutor.scala:
```diff
@@ -0,0 +1,332 @@
+  /** A class to facilitate scheduling and executing a task.
+    *
+    * @param taskRunner the task runner used to execute the task
+    * @param latch the latch that will be zero when the task is scheduled to execute.
```
I thought the TaskExecutor was only handed tasks that have no dependencies.  Why does it need to see this?

nh13 replied:
It may not be dependent on another task, but there also may not be enough resources (currently).

tfenne commented on core/src/main/scala/dagr/core/execsystem2/local/LocalTaskExecutor.scala:
```diff
@@ -0,0 +1,332 @@
+    *                      started executing.
+    * @param resources the resources that the task was scheduled with, None if not scheduled.
+    */
+  protected case class TaskInfo
```
Argh!  The 5th class/trait/object named `TaskInfo`!

tfenne commented on core/src/main/scala/dagr/core/execsystem2/local/LocalTaskExecutor.scala:
```diff
@@ -0,0 +1,332 @@
+  * @param ex the execution context.
+  */
+class LocalTaskExecutor(systemResources: SystemResources = LocalTaskExecutorDefaults.defaultSystemResources,
+                        scriptsDirectory: Option[Path] = None,
```
Why not make this `Path` with a default value?

nh13 replied:
And up the rabbit hole we go with defaults!

tfenne commented on core/src/main/scala/dagr/core/execsystem2/local/LocalTaskExecutor.scala:
```diff
@@ -0,0 +1,332 @@
+  */
+class LocalTaskExecutor(systemResources: SystemResources = LocalTaskExecutorDefaults.defaultSystemResources,
+                        scriptsDirectory: Option[Path] = None,
+                        logDirectory: Option[Path] = None,
```
Ditto.

tfenne commented on core/src/main/scala/dagr/core/execsystem2/local/LocalTaskExecutor.scala:
```diff
@@ -0,0 +1,332 @@
+  }
+
+  // true if the task has been scheduled for execution or is executing.
+  //def scheduled(task: UnitTask): Boolean = this.taskInfo(task).resources.isDefined
```
Is this not needed?

tfenne commented on core/src/main/scala/dagr/core/execsystem2/local/LocalTaskExecutor.scala:
```diff
@@ -0,0 +1,332 @@
+  override def logDir: DirPath = actualLogsDirectory
+
+  /** Returns true if the task is running, false if not running or not tracked */
+  def running(task: UnitTask): Boolean = {
```
`@override`? (here and below)

tfenne commented on core/src/main/scala/dagr/core/execsystem2/local/LocalTaskRunner.scala:
```diff
@@ -0,0 +1,157 @@
+  /** The method to execute the underlying task.  Completes when the underlying task completes. */
+  final def execute()(implicit ex: ExecutionContext): Future[UnitTask] = {
+    val future = blocking { _execute() } interruptable()
+    interruptibleFuture = Some(future)
```
this. please

nh13 commented:
@tfenne I spent a bunch of time yesterday and today going over your comments.  Highlights include:
- attempted to document the main traits and classes
- removed the `DependencyGraph` class as most of its functionality was moved elsewhere
- moved a lot of the methods in `TopLikeStatusReporter` to it's own `dagr.api.models.Executor` API.
- got rid of the need for two separate `Resource` classes
- reorganized the `dagr.api` package (made sub-packages)
I generally followed your comments, and where I didn't, I made sure to respond to your comments.  Therefore, I think it's in good shape for another review (ignore the server and ui projects)
